### DISCUSSION TITLE
Problem in realizing a hexagonal lattice with square or hexagonal holes?

### DISCUSSION THREAD

**Discussion by knifelees3** [NanoComp/meep#2496] [Q&A]
Previously I asked the method for realizing the hexagonal lattice with circle holes, and it can be realized by myself now #2484 However, by changing the holes into other shapes, the permitivity distribution is not I want
![image](https://user-images.githubusercontent.com/18129412/235292013-949bb206-054f-41e1-b10b-b3b87a5e23cf.png)

The following is the program:

```python
import math
import meep as mp
from meep import mpb
import numpy as np
import matplotlib.pyplot as plt

num_bands = 4

# 定义好扫描的k波矢
k_points = [mp.Vector3(),          # Gamma
            mp.Vector3(1/3,2/3),       # K
            mp.Vector3(0.5,0.5),  # M
            mp.Vector3()]          # Gamma

# 对k波矢进行插值
k_points = mp.interpolate(40, k_points)

#  units of nm
a=1



# 基本结构和材料性质
# 折射率设置
nslab = 2.48
slab = mp.Medium(index=nslab)

nair = 1
air = mp.Medium(index=nair)



s=0.25

# square holes
# geometry = [mp.Block(mp.Vector3(0.2,0.2),center=mp.Vector3(-0.25/2,-np.sqrt(3)/4/2), material=air),
#            mp.Block(mp.Vector3(0.2,0.2),center=mp.Vector3(0.25/2,np.sqrt(3)/4/2), material=air)]

## hexagonal holes
s1=0.1
s2=0.2
vertices1 = [mp.Vector3(-1*s1,0),
            mp.Vector3(-0.5*s1,math.sqrt(3)/2*s1),
            mp.Vector3(0.5*s1,math.sqrt(3)/2*s1),
            mp.Vector3(1*s1,0),
            mp.Vector3(0.5*s1,-math.sqrt(3)/2*s1),
            mp.Vector3(-0.5*s1,-math.sqrt(3)/2*s1)]

vertices2 = [mp.Vector3(-1*s2,0),
            mp.Vector3(-0.5*s2,math.sqrt(3)/2*s2),
            mp.Vector3(0.5*s2,math.sqrt(3)/2*s2),
            mp.Vector3(1*s2,0),
            mp.Vector3(0.5*s2,-math.sqrt(3)/2*s2),
            mp.Vector3(-0.5*s2,-math.sqrt(3)/2*s2)]
geometry = [mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=air),
            
           mp.Prism(vertices2, height=mp.inf, center=mp.Vector3(0,np.sqrt(3)/4),material=air)]



geometry_lattice = mp.Lattice(
    size=mp.Vector3(1, 1),
    basis1=mp.Vector3(0.5, np.sqrt(3) / 2),
    basis2=mp.Vector3(1,0),
)


resolution = 32

# mode solver设置
ms = mpb.ModeSolver(num_bands=num_bands,
                    k_points=k_points,
                    geometry=geometry,
                    geometry_lattice=geometry_lattice,
                    default_material=slab,
                    resolution=resolution)

ms.run_te()
te_freqs = ms.all_freqs
te_gaps = ms.gap_list
```
![image](https://user-images.githubusercontent.com/18129412/235292149-bc6a4bda-2ef5-4db2-93f6-5dce1ff8bb38.png)
![image](https://user-images.githubusercontent.com/18129412/235292329-701f3dc1-a1cd-4827-bc87-4f5580ce8eae.png)

Do I have to pre-warp the shape inside the cell in order to get the setup right? For example, a square is deformed and transformed into a corresponding rhombus, so that the final result is a square? I also don't understand why there is no problem with the round shape.

Or is it just a matter of MEEP display? Calculations are not affected?

**Comment by stevengj** [comment]
> For example, a square is deformed and transformed into a corresponding rhombus

See the second paragraph in [units](https://mpb.readthedocs.io/en/latest/Python_Tutorial/#a-few-words-on-units) — the MPB coordinates are **not Cartesian**.  They are in the basis of the lattice vectors.

If you want to specify your `Prism` in Cartesian coordinates, that's fine!  Just `map` the vertex list through [`cartesian_to_lattice`](https://mpb.readthedocs.io/en/latest/Python_User_Interface/#coordinate-conversion-functions) before passing them to the `Prism` constructor.

Same thing with a `Block` object — by default, its axes are `e1 = (1,0,0)` and `e2 = (0,1,0)` *in the lattice basis*, i.e. the default `Block` is a rhombus aligned with the lattice.   If you want a square, you should instead pass `e1 = cartesian_to_lattice((0,1,0), geometry_lattice),   e2 = cartesian_to_lattice((1,1,0), geometry_lattice)` to the `Block` constructor.

**Reply by knifelees3** [reply]
Thanks for you detailed explanations and I can now reproduce more complex structures, Especially the periodic structure below contains six rods, 

![image](https://user-images.githubusercontent.com/18129412/235616047-8c1c640a-6511-43a5-9393-6a6cf6288fa7.png)

because the overall size of the six rods has to shrink at will, so the primitive cell cannot only contain two pillars. Through coordinate transformation function [cartesian_to_lattice](https://mpb.readthedocs.io/en/latest/Python_User_Interface/#coordinate-conversion-functions), I can already get the permitivity distribution I want very correctly. The simulation result is completely consistent with the original [paper](https://www.google.com/search?q=scheme+for+achieving+a+topological+photonic+crystal+by+using+dielectric+material&oq=Scheme+for+Achieving+a+Topological+Photonic+Crystal+by+Using+Dielectric+Material&aqs=chrome.0.35i39i650.752j0j1&sourceid=chrome&ie=UTF-8),  Detailed program can be found in this [notebook](https://github.com/knifelees3/MEEPStudy/blob/master/MEEPBandStructuresSimulations_2DTopological.ipynb)

I can also realize the square holes by changing the axes `e1,e2`:
```python
s=0.125
geometry=[mp.Block(mp.Vector3(0.25,0.25),center=mp.Vector3(0,0), material=slab,
                  e1 = mp.cartesian_to_lattice(mp.Vector3(1,0,0), geometry_lattice),
                  e2 = mp.cartesian_to_lattice(mp.Vector3(0,1,0), geometry_lattice))]
```
![image](https://user-images.githubusercontent.com/18129412/235679196-7870ee9e-2783-4684-b1f4-f500ddd12ae4.png)

but it fails by using the `Prism` method, for the sake of intuitive illustration, I directly use the method of  `Prism` to implement Square holes, by using cartesian_to_lattice
```python
s=0.125
vertices1 = [mp.cartesian_to_lattice(mp.Vector3(-1*s,-1*s),geometry_lattice),
             mp.cartesian_to_lattice(mp.Vector3(-1*s,1*s),geometry_lattice),
             mp.cartesian_to_lattice(mp.Vector3(s*1,s*1),geometry_lattice),
             mp.cartesian_to_lattice(mp.Vector3(1*s,-1*s),geometry_lattice)]
geometry=[mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=slab)]
```
or directly passing the coordinates in lattice coordinates
```python
 s=0.2
vertices1 = [mp.Vector3(0*s,0*s),
             mp.Vector3(1*s,0*s),
              mp.Vector3(s*3/2,s*np.sqrt(3)/2),
              mp.Vector3(s*1/2,s*np.sqrt(3)/2),]
geometry=[mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=slab)]
```

without rectify
![image](https://user-images.githubusercontent.com/18129412/235680221-5e73bc3a-d4c3-4bea-a4c5-e2dc41ed486f.png)

with rectify
![image](https://user-images.githubusercontent.com/18129412/235680273-d6cbfda6-c253-4847-b59c-6e9dc5a6f88a.png)

It seems that by using   `Prism`, the corresponding coordinates are not correctly assigned in lattice coordinates. The full program is as follows
```python
# we first use the function mp.cartesian_to_lattice to transform the coordinates

import math

import meep as mp
from meep import mpb
import numpy as np
# A triangular lattice of dielectric rods in air.  (This structure has
# a band_gap for TM fields.)  This file is used in the "Data Analysis
# Tutorial" section of the MPB manual.

num_bands = 8


nslab = 2.48
slab = mp.Medium(index=nslab)


# Lattice geometry
geometry_lattice = mp.Lattice(
    size=mp.Vector3(1, 1),
    basis2=mp.Vector3(0.5,math.sqrt(3) / 2),
    basis1=mp.Vector3(1, 0),
)

# (1) For squares in cartesian coordinates
# s=0.125
# vertices1 = [mp.Vector3(-1*s,-1*s),
#              mp.Vector3(-1*s,1*s),
#              mp.Vector3(s*1,s*1),
#              mp.Vector3(1*s,-1*s)]
# geometry=[mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=slab)]



#(2) For squares in Lattice coordinates method 1
# s=0.125
# vertices1 = [mp.cartesian_to_lattice(mp.Vector3(-1*s,-1*s),geometry_lattice),
#              mp.cartesian_to_lattice(mp.Vector3(-1*s,1*s),geometry_lattice),
#              mp.cartesian_to_lattice(mp.Vector3(s*1,s*1),geometry_lattice),
#              mp.cartesian_to_lattice(mp.Vector3(1*s,-1*s),geometry_lattice)]
# geometry=[mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=slab)]

# # (3) For squares in Lattice coordinates method 2
# s=0.125
# geometry=[mp.Block(mp.Vector3(0.25,0.25),center=mp.Vector3(0,0), material=slab)]

# #(4) For squares in Lattice coordinates method 2
# s=0.125
# geometry=[mp.Block(mp.Vector3(0.25,0.25),center=mp.Vector3(0,0), material=slab,
#                   e1 = mp.cartesian_to_lattice(mp.Vector3(1,0,0), geometry_lattice),
#                   e2 = mp.cartesian_to_lattice(mp.Vector3(0,1,0), geometry_lattice))]

# # #(5) For squares in Lattice coordinates method 3
# s=0.2
# vertices1 = [mp.Vector3(0*s,0*s),
#              mp.Vector3(1*s,0*s),
#              mp.Vector3(s*3/2,s*np.sqrt(3)/2),
#              mp.Vector3(s*1/2,s*np.sqrt(3)/2),]
# geometry=[mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=slab)]


# 注意倒格矢空间
k_points = [
    mp.Vector3(),  # Gamma
    mp.Vector3(y=0.5),  # M
    mp.Vector3(1 / -3, 1 / 3),  # K
    mp.Vector3(),  # Gamma
]

k_points = mp.interpolate(4, k_points)

resolution = 32

ms = mpb.ModeSolver(
    geometry=geometry,
    geometry_lattice=geometry_lattice,
    k_points=k_points,
    resolution=resolution,
    num_bands=num_bands,
)

ms.run_tm()
tm_freqs = ms.all_freqs
tm_gaps = ms.gap_list

import matplotlib.pyplot as plt
md = mpb.MPBData(rectify=False, periods=6, resolution=64)
eps = ms.get_epsilon()
converted_eps = md.convert(eps)

plt.imshow(converted_eps.T, interpolation='spline36', cmap='binary')
plt.axis('off')
plt.show()

md = mpb.MPBData(rectify=True, periods=6, resolution=64)
eps = ms.get_epsilon()
converted_eps = md.convert(eps)

plt.imshow(converted_eps.T, interpolation='spline36', cmap='binary')
plt.axis('off')
plt.show()
```
